File: examples/test_csr.cpp
/*
   Register access functions for RISC-V system registers.
   SPDX-License-Identifier: Unlicense

   https://five-embeddev.com/

*/

#include "riscv-csr.hpp"

static volatile riscv::csr::uint_xlen_t saved_mtvec;

int main(void) {

    riscv::csr::all csrs;

    auto cause = csrs.mcause.read();
    if (cause & riscv::csr::mcause_data::interrupt::BIT_MASK) {
        // Interrupt
        auto mip = csrs.mip.read();
    } else {
        // Exception
        // Read the source of the exception
        auto mepc = csrs.mepc.read();
        // Pointless save to mscratch
        csrs.mscratch.write(mepc);
    }
    // Enable MIE.MSI    
    csrs.mie.msi.set();
    // Enable MIE.MTI
    csrs.mie.mti.set();
    // Clear mode and read mtvec.
    auto old_mtvec = csrs.mtvec.read_clr_bits_const<riscv::csr::mtvec_data::mode::ALL_SET_MASK>();
    if (old_mtvec & riscv::csr::mtvec_data::mode::ALL_SET_MASK) {
        // Remove mode bits
        old_mtvec &= riscv::csr::mtvec_data::base::ALL_SET_MASK;
    }
    saved_mtvec = old_mtvec;

    // ----
    // Try writing const values

    // Write via a register value to the MTVEC
    csrs.mtvec.write(0x10);
    // Write an immediate value to the MTVEC
    csrs.mtvec.write_const<0x10>();
    // Write via a register value to the MTVEC
    csrs.mtvec.write_const<0x100000>();

    // ----
    // Try out the operators

    // Read mcycle via default operator
    auto mcycle = csrs.mcycle();

#if 0 // TODO - upgrade to gcc that supports these CSRs
    // Disable all the counters
    csrs.mcountinhibit |= 0xFFFFFFFF;

    // Enable the instruction retired count
    csrs.mcountinhibit.ir.clr();
    // Enable the cycle count
    csrs.mcountinhibit.cy.write(0);
    // Enable the hpm count
    csrs.mcountinhibit.hpm.clr();
#endif
}

File: rs/riscv_csr_macros/examples/test_csr.rs
/*
   Register access functions for RISC-V system registers.
   SPDX-License-Identifier: Unlicense

   https://five-embeddev.com/

*/
#![no_std]
#![no_main]

use core::panic::PanicInfo;

// error: `#[panic_handler]` function required, but not found
#[panic_handler]
fn panic(_info: &PanicInfo) -> ! {

    loop {}
}

use riscv_csr_macros::csr_read_mcause;
use riscv_csr_macros::csr_read_mip;
use riscv_csr_macros::csr_read_mepc;
use riscv_csr_macros::csr_write_mscratch;
use riscv_csr_macros::csr_read_clr_bits_mtvec;
use riscv_csr_macros::csr_set_bits_mie;
use riscv_csr_macros::csr_set_bits_imm_mie;
use riscv_csr_macros::MCAUSE_INTERRUPT_BIT_MASK;
use riscv_csr_macros::MIE_MTI_BIT_MASK;
use riscv_csr_macros::MTVEC_BASE_ALL_SET_MASK;
use riscv_csr_macros::MTVEC_MODE_ALL_SET_MASK;
use riscv_csr_macros::UintXlen;


#[no_mangle]
pub fn test_csr() {
    let mut _saved_mtvec: UintXlen = 0;
    let cause = csr_read_mcause!();
    if (cause & MCAUSE_INTERRUPT_BIT_MASK) != 0 {
        // Interrupt
        let _mip = csr_read_mip!();
    } else {
        // Exception
        // Read the source of the exception
        let mepc =  csr_read_mepc!();
        // Pointless save to mscratch
        csr_write_mscratch!(mepc);
    }
    // Enable MIE.MSI    
    csr_set_bits_imm_mie!(MIE_MSI_BIT_MASK);
    // Enable MIE.MTI
    csr_set_bits_mie!(MIE_MTI_BIT_MASK);
    // Clear mode and read mtvec.
    let mut old_mtvec = csr_read_clr_bits_mtvec!(MTVEC_MODE_ALL_SET_MASK);
    if (old_mtvec & MTVEC_MODE_ALL_SET_MASK) != 0 {
        // Remove mode bits
        old_mtvec = old_mtvec & MTVEC_BASE_ALL_SET_MASK;
    }
    _saved_mtvec = old_mtvec;
}
